autoPyTorch/pipeline/components/training/trainer/base_trainer.py

```diff
@@ -0,0 +1,351 @@
+        self.device = device
+
+        # Setup the metrics
+        self.metrics = metrics
+
+        # Setup the loss function
+        self.criterion = criterion.to(device)
+
+        # setup the model
+        self.model = model.to(device)
+
+        # setup the optimizers
+        self.optimizer = optimizer
+
+        # The budget tracker
+        self.budget_tracker = budget_tracker
+
+        # A summary writer for tensorboard
+        self.writer = writer
+
+    def on_epoch_start(self, X: Dict[str, Any], epoch: int) -> None:
+        """
+        Optional place holder for AutoPytorch Extensions.
+
+        An user can define what happens on every epoch start or every epoch end.
+        """
+        pass
+
+    def on_epoch_end(self, X: Dict[str, Any], epoch: int) -> bool:
+        """
+        Optional place holder for AutoPytorch Extensions.
+        An user can define what happens on every epoch start or every epoch end.
+        If returns True, the training is stopped
+
+        """
+        return False
+
+    def train(self, train_loader: torch.utils.data.DataLoader, epoch: int
+              ) -> Tuple[float, Dict[str, float]]:
+        '''
+            Trains the model for a single epoch.
+
+        Args:
+            train_loader (torch.utils.data.DataLoader): generator of features/label
+            epoch (int): The current epoch used solely for tracking purposes
+
+        Returns:
+            float: training loss
+            Dict[str, float]: scores for each desired metric
+        '''
+
+        loss_sum = 0.0
+        N = 0
+        self.model.train()
+        outputs_data = list()
+        targets_data = list()
+
+        for step, (data, targets) in enumerate(train_loader):
+
+            if self.budget_tracker.is_max_time_reached():
+                self.logger.info("Stopping training as max time reached")
+                break
+
+            # prepare
+            data = data.float().to(self.device)
+            targets = targets.long().to(self.device)
+
+            data, criterion_kwargs = self.data_preparation(data, targets)
+            data = Variable(data)
+            batch_size = data.size(0)
+
+            # training
+            self.optimizer.zero_grad()
+            outputs = self.model(data)
+            loss_func = self.criterion_preparation(**criterion_kwargs)
+            loss = loss_func(self.criterion, outputs)
+            loss.backward()
+            self.optimizer.step()
+
+            # save for metric evaluation
+            outputs_data.append(outputs.detach())
+            targets_data.append(targets.detach())
+
+            loss_sum += loss.item() * batch_size
+            N += batch_size
+
+            if self.writer:
+                self.writer.add_scalar(
+                    'Train/loss',
+                    loss.item(),
+                    epoch * len(train_loader) + step,
+                )
+
+        return loss_sum / N, self.compute_metrics(outputs_data, targets_data)
```

Comment:
I think we should add an option to not evaluate train metrics for better anytime performance

Reply:
Done in the new push!